### ISSUE TITLE
Hash sums don't work in rclone serve sftp preventing file-creating operations effectively forcing the use of --sftp-disable-hashcheck.

### ISSUE THREAD

**Issue by pb2004** [rclone/rclone#7241]
#### What is the problem you are having with rclone?
rclone serve sftp does not create checksums but returns the error 'command "md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial" failed with error: unexpected non file'. This makes it impossible, for example, to copy a new file.

#### What is your rclone version (output from `rclone version`)
**rclone serve sftp:**
```
rclone v1.63.1
- os/version: debian 12.1 (64 bit)
- os/kernel: 6.1.0-11-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.20.6
- go/linking: static
- go/tags: none
```
**rclone copyto:**
```
rclone v1.63.1
- os/version: Microsoft Windows 10 Pro 22H2 (64 bit)
- os/kernel: 10.0.19045.3324 (x86_64)
- os/type: windows
- os/arch: amd64
- go/version: go1.20.6
- go/linking: static
- go/tags: cmount
```

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
os/version: debian 12.1 (64 bit)
and
os/version: Microsoft Windows 10 Pro 22H2 (64 bit)

#### Which cloud storage system are you using? (e.g. Google Drive)
sftp

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
On server:
`rclone serve sftp / --addr :2022 --vfs-cache-mode full --authorized-keys "~/.ssh/authorized_keys" -vv -L`
on client:
`rclone.exe copyto M:\gocryptfs\usAk8EPwgCbWkHVJgWDrjQ sftpvdbtr2:home\piotr\usAk8EPwgCbWkHVJgWDrjQ -vv
`
configuration of sftpvdbtr2:
```
[sftpvdbtr2]
type = sftp
host = vdebtor
user = piotr
port = 2022
key_use_agent = true
shell_type = unix
md5sum_command = md5sum
sha1sum_command = sha1sum
```
#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
Relevant part of serve sftp:

```
2023/08/20 20:54:54 INFO  : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: vfs cache: queuing for upload in 5s
2023/08/20 20:54:54 DEBUG : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial(0xc00028a600): >close: err=<nil>
2023/08/20 20:54:54 DEBUG : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: vfs cache: setting modification time to 2023-08-19 13:47:32 +0200 CEST
2023/08/20 20:54:54 DEBUG : serve sftp [fe80::ac3b:54d8:6f83:31fb%ens33]:61727->[fe80::e30f:27e2:86a3:1619%ens33]:2022: Incoming channel: session
2023/08/20 20:54:54 DEBUG : serve sftp [fe80::ac3b:54d8:6f83:31fb%ens33]:61727->[fe80::e30f:27e2:86a3:1619%ens33]:2022: Channel accepted
2023/08/20 20:54:54 DEBUG : serve sftp [fe80::ac3b:54d8:6f83:31fb%ens33]:61727->[fe80::e30f:27e2:86a3:1619%ens33]:2022: Request: exec
2023/08/20 20:54:54 DEBUG : serve sftp [fe80::ac3b:54d8:6f83:31fb%ens33]:61727->[fe80::e30f:27e2:86a3:1619%ens33]:2022:  - accepted: true
2023/08/20 20:54:54 DEBUG : serve sftp [fe80::ac3b:54d8:6f83:31fb%ens33]:61727->[fe80::e30f:27e2:86a3:1619%ens33]:2022: exec command: binary = "md5sum", args = "/home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial"
2023/08/20 20:54:54 DEBUG : serve sftp [fe80::ac3b:54d8:6f83:31fb%ens33]:61727->[fe80::e30f:27e2:86a3:1619%ens33]:2022: command "md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial" failed with error: unexpected non file
2023/08/20 20:54:54 DEBUG : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: Remove: 
2023/08/20 20:54:54 DEBUG : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: vfs cache: cancelling writeback (uploading false) 0xc0002901c0 item 6
2023/08/20 20:54:54 INFO  : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: vfs cache: removed cache file as file deleted
2023/08/20 20:54:54 DEBUG : home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: vfs cache: removed metadata from cache as file deleted
```

Relevant part of copyto:

```
2023/08/20 20:54:53 DEBUG : usAk8EPwgCbWkHVJgWDrjQ: Need to transfer - File not found at Destination
2023/08/20 20:54:53 DEBUG : sftp://piotr@vdebtor:2022/home/piotr/: Shell path "/home/piotr/usAk8EPwgCbWkHVJgWDrjQ.lefusif1.partial"
2023/08/20 20:54:53 DEBUG : sftp://piotr@vdebtor:2022/home/piotr/: Running remote command: md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.lefusif1.partial
2023/08/20 20:54:53 ERROR : usAk8EPwgCbWkHVJgWDrjQ.lefusif1.partial: Failed to calculate dst hash: failed to calculate md5 hash: failed to run "md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.lefusif1.partial": unexpected non file: Process exited with status 1
2023/08/20 20:54:53 ERROR : usAk8EPwgCbWkHVJgWDrjQ.lefusif1.partial: corrupted on transfer: md5 hash differ "c6b0e389f36ab712c554c56c555f2665" vs ""
2023/08/20 20:54:53 INFO  : usAk8EPwgCbWkHVJgWDrjQ.lefusif1.partial: Removing failed copy
2023/08/20 20:54:53 ERROR : Attempt 1/3 failed with 2 errors and: corrupted on transfer: md5 hash differ "c6b0e389f36ab712c554c56c555f2665" vs ""
2023/08/20 20:54:53 DEBUG : usAk8EPwgCbWkHVJgWDrjQ: Need to transfer - File not found at Destination
2023/08/20 20:54:53 DEBUG : sftp://piotr@vdebtor:2022/home/piotr/: Shell path "/home/piotr/usAk8EPwgCbWkHVJgWDrjQ.nopagok3.partial"
2023/08/20 20:54:53 DEBUG : sftp://piotr@vdebtor:2022/home/piotr/: Running remote command: md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.nopagok3.partial
2023/08/20 20:54:53 ERROR : usAk8EPwgCbWkHVJgWDrjQ.nopagok3.partial: Failed to calculate dst hash: failed to calculate md5 hash: failed to run "md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.nopagok3.partial": unexpected non file: Process exited with status 1
2023/08/20 20:54:53 ERROR : usAk8EPwgCbWkHVJgWDrjQ.nopagok3.partial: corrupted on transfer: md5 hash differ "c6b0e389f36ab712c554c56c555f2665" vs ""
2023/08/20 20:54:53 INFO  : usAk8EPwgCbWkHVJgWDrjQ.nopagok3.partial: Removing failed copy
2023/08/20 20:54:53 ERROR : Attempt 2/3 failed with 2 errors and: corrupted on transfer: md5 hash differ "c6b0e389f36ab712c554c56c555f2665" vs ""
2023/08/20 20:54:53 DEBUG : usAk8EPwgCbWkHVJgWDrjQ: Need to transfer - File not found at Destination
2023/08/20 20:54:54 DEBUG : sftp://piotr@vdebtor:2022/home/piotr/: Shell path "/home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial"
2023/08/20 20:54:54 DEBUG : sftp://piotr@vdebtor:2022/home/piotr/: Running remote command: md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial
2023/08/20 20:54:54 ERROR : usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: Failed to calculate dst hash: failed to calculate md5 hash: failed to run "md5sum /home/piotr/usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial": unexpected non file: Process exited with status 1
2023/08/20 20:54:54 ERROR : usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: corrupted on transfer: md5 hash differ "c6b0e389f36ab712c554c56c555f2665" vs ""
2023/08/20 20:54:54 INFO  : usAk8EPwgCbWkHVJgWDrjQ.wukuvut8.partial: Removing failed copy
2023/08/20 20:54:54 ERROR : Attempt 3/3 failed with 2 errors and: corrupted on transfer: md5 hash differ "c6b0e389f36ab712c554c56c555f2665" vs ""
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
I've managed to replicate this.

The cause is the `--vfs-cache-mode full` - if you use `off` then it works just fine.

The problem is that we should be using the VFS layer to read the MD5 sum here

Give this a go

[v1.64.0-beta.7252.270b86737.fix-serve-sftp](https://beta.rclone.org/branch/fix-serve-sftp/v1.64.0-beta.7252.270b86737.fix-serve-sftp/) on branch [fix-serve-sftp](https://github.com/rclone/rclone/tree/fix-serve-sftp) (uploaded in 15-30 mins)

**Comment by pb2004** [comment]
> I've managed to replicate this.
> 
> The cause is the `--vfs-cache-mode full` - if you use `off` then it works just fine.
> 
> The problem is that we should be using the VFS layer to read the MD5 sum here
> 
> Give this a go
> 
> [v1.64.0-beta.7252.270b86737.fix-serve-sftp](https://beta.rclone.org/branch/fix-serve-sftp/v1.64.0-beta.7252.270b86737.fix-serve-sftp/) on branch [fix-serve-sftp](https://github.com/rclone/rclone/tree/fix-serve-sftp) (uploaded in 15-30 mins)

It now works as expected. Thanks.

**Comment by ncw** [comment]
Thank you for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.64
